OpenSim/Simulation/Model/MovingPathPoint.cpp

```diff
@@ -47,472 +43,337 @@ using SimTK::Vec3;
 {
-    if (_xCoordinate) {
+    if (!_xCoordinate.empty()) {
         // If the function is already a MultiplierFunction, just update its scale factor.
         // Otherwise, make a MultiplierFunction from it and make the muscle point use
         // the new MultiplierFunction.
-        MultiplierFunction* mf = dynamic_cast<MultiplierFunction*>(_xLocation);
+        MultiplierFunction* mf = dynamic_cast<MultiplierFunction*>(&upd_x_location());
         if (mf) {
             mf->setScale(mf->getScale() * aScaleFactors[0]);
         } else {
             // Make a copy of the original function and delete the original
             // (so its node will be removed from the XML document).
-            mf = new MultiplierFunction(_xLocation->clone(), aScaleFactors[0]);
-            delete _xLocation;
-            setXFunction(s, *mf);
+            set_x_location(MultiplierFunction(get_x_location().clone(), aScaleFactors[0]));
         }
     }
 
-    if (_yCoordinate) {
+    if (!_yCoordinate.empty()) {
         // If the function is already a MultiplierFunction, just update its scale factor.
         // Otherwise, make a MultiplierFunction from it and make the muscle point use
         // the new MultiplierFunction.
-        MultiplierFunction* mf = dynamic_cast<MultiplierFunction*>(_yLocation);
+        MultiplierFunction* mf = dynamic_cast<MultiplierFunction*>(&upd_y_location());
         if (mf) {
             mf->setScale(mf->getScale() * aScaleFactors[1]);
         } else {
             // Make a copy of the original function and delete the original
             // (so its node will be removed from the XML document).
-            mf = new MultiplierFunction(_yLocation->clone(), aScaleFactors[1]);
-            delete _yLocation;
-            setYFunction(s, *mf);
+            set_y_location(MultiplierFunction(get_y_location().clone(), aScaleFactors[1]));
         }
     }
 
-    if (_zCoordinate) {
+    if (!_zCoordinate.empty()) {
         // If the function is already a MultiplierFunction, just update its scale factor.
         // Otherwise, make a MultiplierFunction from it and make the muscle point use
         // the new MultiplierFunction.
-        MultiplierFunction* mf = dynamic_cast<MultiplierFunction*>(_zLocation);
+        MultiplierFunction* mf = dynamic_cast<MultiplierFunction*>(&upd_z_location());
         if (mf) {
             mf->setScale(mf->getScale() * aScaleFactors[2]);
         } else {
-            mf = new MultiplierFunction(_zLocation->clone(), aScaleFactors[2]);
-            delete _zLocation;
-            setZFunction(s, *mf);
+            set_z_location(MultiplierFunction(get_z_location().clone(), aScaleFactors[2]));
         }
     }
 
     updateGeometry();
 }
+
+
+SimTK::Vec3 MovingPathPoint::calcLocationInGround(const SimTK::State& s) const
+{
+    return getParentFrame().getTransformInGround(s)*getLocation(s);
+}
+
+SimTK::Vec3 MovingPathPoint::calcVelocityInGround(const SimTK::State& s) const
```

Comment:
I did not include additional unit tests. `testMomentArms` was my working test case for checking my implementation was correct (or at least produces consistent moment-arms).  These functions are also required to evaluate muscle path lengths and speeds (for models with MovingPathPoints) which are cases in `testForward`, `testStaticOptimization`, `testCMC`, etc...  

My intention is to eliminate `MovingPathPoint`, and have all Points provide their Jacobian with respect to any set of Coordinates. The implementation of `MovingPathPoint` makes little sense with exactly one Coordinate assigned to each component of the location vector. Once points can provide their partials I plan to test that more extensively.